src/main.rs

```diff
@@ -320,13 +357,96 @@ async fn produce_peer_events(
     }
 }
 
+#[async_trait]
+trait IncomingMessageProducer {
+    async fn receive(&mut self) -> Result<CustomMessage, Status>;
+}
+
+struct MessageStream {
+    message_subscription: tonic_lnd::tonic::Streaming<CustomMessage>,
+}
+
+#[async_trait]
+impl IncomingMessageProducer for MessageStream {
+    async fn receive(&mut self) -> Result<CustomMessage, Status> {
+        match self.message_subscription.message().await? {
+            Some(msg) => Ok(msg),
+            None => Err(Status::unknown("no message provided")),
+        }
+    }
+}
+
+/// Consumes a stream of incoming message events from the IncomingMessageProducer until the stream exits (by sending an
+/// error) or the producer receives the signal to exit (via close of the exit channel).
+///
+/// Note that this function *must* send an exit error to the Sender provided on all exit-cases, so that upstream
+/// consumers know to exit as well. Failures related to sending events are an exception, as failure to send indicates
+/// that the consumer has already exited (the receiving end of the channel has hung up), and we can't send any more
+/// events anyway.
+async fn produce_incoming_message_events(
+    mut source: impl IncomingMessageProducer,
+    events: Sender<MessengerEvents>,
+    mut exit: Receiver<()>,
+) -> Result<(), ProducerError> {
+    loop {
+        select! (
+        // Select biased so that we'll always check our exit signal before attempting to receive. This allows more
+        // deterministic tests, and ensures that the producer will exit when requested (and won't queue up a series
+        // of events that can't be consumed, possibly blocking if the channel buffer is small).
+        biased;
+
+        _ = exit.recv() => {
+            info!("Peer events received signal to quit");
+            return Ok(())
+        }
+        onion_message = source.receive() => {
+            match onion_message {
+                Ok(incoming_message) => {
+                    if incoming_message.r#type != ONION_MESSAGE_TYPE {
+                        trace!("Ignoring custom message: {}", incoming_message.r#type);
+                        continue;
+                    }
+
+                    let pubkey = PublicKey::from_slice(&incoming_message.peer).unwrap();
```

Comment:
Instead of unwrap, would it make sense to log something if this does throw an error?

Reply:
Hm, on the fence about this one - i:f LND gives us an unparsable pubkey then something has gone _really_ wrong, so I think it's ok to have a messy exit?

Reply:
Hmmm yeah that's fair. I guess I'm thinking about a possible case where for some reason we're getting bad data for one of the peers, but not in future messages, it would be nice to handle that with more resilience. But maybe that's unlikely.

Reply:
In this case the pubkey is provided by LND's API, so it'll be safe. But def agree on peer provdied data - just log and move on - doing that [below](https://github.com/carlaKC/lndk/pull/29/files?diff=unified&w=0#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR421) if we get an onion message we can't parse. Is that the sort of case you were thinking about?